Havit.Blazor.Components.Web.Bootstrap/Forms/Internal/HxMultiSelectInternal.razor.cs

```diff
@@ -49,36 +53,252 @@ public partial class HxMultiSelectInternal<TValue, TItem>
+			{
+				// We don't want to trigger select all triggers each time, an item is selected, just once at the end
+				await HandleItemSelectionChangedAsync(false, item, false);
+			}
+		}
+		else
+		{
+			foreach (var item in filteredItems)
+			{
+				var value = SelectorHelpers.GetValue<TItem, TValue>(ValueSelector, item);
+				var itemSelected = DoSelectedValuesContainValue(value);
+
+				// If the item is already selected we don't need to reselect it
+				if (!itemSelected)
+				{
+					// We don't want to trigger select all triggers each time, an item is selected, just once at the end
+					await HandleItemSelectionChangedAsync(!itemSelected, item, false);
+				}
+			}
+		}
+
+		await ChangeSelectAllAsync(!selectAll);
+	}
+
+	private bool DoSelectedValuesContainValue(TValue value)
+	{
+		return SelectedValues?.Contains(value) ?? false;
+	}
+
+	private Task ChangeSelectAllAsync(bool selectAll)
+	{
+		this.selectAll = selectAll;
+		return SelectAllChanged.InvokeAsync(selectAll);
+	}
+
+	private Task HandleInputChangedAsync(ChangeEventArgs e)
+	{
+		filterText = e.Value?.ToString() ?? string.Empty;
+		return ChangeSelectAllAsync(false);
+	}
+
+	private List<TItem> GetFilteredItems()
+	{
+		if (!AllowFiltering || string.IsNullOrEmpty(filterText))
+		{
+			return ItemsToRender;
+		}
+
+		var filterPredicate = FilterPredicate ?? DefaultFilterPredicate;
+		return ItemsToRender.Where(x => filterPredicate(x, filterText)).ToList();
+
+		bool DefaultFilterPredicate(TItem item, string filter)
+		{
+			return string.IsNullOrEmpty(filter) || TextSelector(item).Contains(filter, StringComparison.OrdinalIgnoreCase);
+		}
 	}
 
 	public async ValueTask FocusAsync()
 	{
-		if (EqualityComparer<ElementReference>.Default.Equals(inputElement, default))
+		if (EqualityComparer<ElementReference>.Default.Equals(elementReference, default))
 		{
 			throw new InvalidOperationException($"Cannot focus {this.GetType()}. The method must be called after first render.");
 		}
-		await inputElement.FocusAsync();
+		await elementReference.FocusAsync();
+	}
+
+	/// <summary>
+	/// Receives notification from JavaScript when item is hidden.
+	/// </summary>
+	[JSInvokable("HxMultiSelect_HandleJsHidden")]
+	public Task HandleJsHidden()
+	{
+		isShown = false;
+
+		if (ClearFilterOnHide)
+		{
+			filterText = string.Empty;
+		}
+
+		return InvokeOnHiddenAsync(this.InputId);
+	}
+
+	[JSInvokable("HxMultiSelect_HandleJsShown")]
+	public async Task HandleJsShown()
+	{
+		isShown = true;
+		await filterInputReference.FocusAsync();
+		await InvokeOnShownAsync(this.InputId);
+	}
+
+	/// <summary>
+	/// Collapses the item.
+	/// </summary>
+	public async Task HideAsync()
+	{
+		await EnsureJsModuleAsync();
+		await jsModule.InvokeVoidAsync("hide", elementReference);
+	}
+
+	/// <summary>
+	/// Expands the item.
+	/// </summary>
+	public async Task ShowAsync()
+	{
+		await EnsureJsModuleAsync();
+		await jsModule.InvokeVoidAsync("show", elementReference);
+	}
```

Comment:
Please remove the methods. They are not being used nor exposed to the user.

Reply:
Have removed these methods.